sig-auth/cluster-role-aggregation.md

```diff
@@ -0,0 +1,94 @@
+# Cluster Role Aggregation
+In order to support easy RBAC integration for CustomResources and Extension
+APIServers, we need to have a way for API extenders to add permissions to the
+"normal" roles for admin, edit, and view.
+
+These roles express an intent for the namespaced power of administrators of the
+namespace (manage ownership), editors of the namespace (manage content like
+pods), and viewers of the namespace (see what is present).  As new APIs are
+made available, these roles should reflect that intent to prevent migration
+concerns every time a new API is added.
+
+To do this, we will allow one ClusterRole to be built out of a selected set of
+ClusterRoles.
+
+## API Changes
+```yaml
+aggregationRule:
+  selectors:
+  - matchLabels:
+      rbac.authorization.k8s.io/aggregate-to-clusterrole-name: admin
+```
+
+```go
+// ClusterRole is a cluster level, logical grouping of PolicyRules that can be referenced as a unit by a RoleBinding or ClusterRoleBinding.
+type ClusterRole struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	metav1.ObjectMeta
+
+	// Rules holds all the PolicyRules for this ClusterRole
+	Rules []PolicyRule
+
+	// AggregationRule is an optional field that describes how to build the Rules for this ClusterRole.
+	// If AggregationRule is set, then the Rules are controller managed and direct changes to Rules will be
+	// stomped by the controller.
+	AggregationRule *AggregationRule
```

Comment:
> what would reconciliation for this field look like? that will be important in transitioning the existing admin/edit/view roles to aggregated roles

I replaced as a unit: https://github.com/kubernetes/kubernetes/pull/54005/files#diff-769f1c7c52943946bdc1ec3d7c9ec5d2R205 .  It could tighten.  I could write a one-time migration for those three using the reconcile post-start hook if you're particularly concerned.  I think its unlikely that many people will have problems going forward.

Reply:
> what would reconciliation for this field look like? that will be important in transitioning the existing admin/edit/view roles to aggregated roles
> I replaced as a unit: https://github.com/kubernetes/kubernetes/pull/54005/files#diff-769f1c7c52943946bdc1ec3d7c9ec5d2R205 . It could tighten. I could write a one-time migration for those three using the reconcile post-start hook if you're particularly concerned. I think its unlikely that many people will have problems going forward.

I have updated this so that selectors which are not present in the actual get added.  On a tightening reconcile, the extra selectors are removed.